### ISSUE TITLE
getc() and ungetc() with unicode failure

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12691]
Migrated from [rt.perl.org#116322](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=116322) (status was 'resolved')

Searchable as RT116322$

**Comment by p5pRT** [comment]
### From michael@negativespace.net

#### Created by michael@negativespace.net

This is a bug report for perl from michael@&#8203;negativespace\.net\,
generated with the help of perlbug 1\.39 running under perl 5\.16\.2\.


getc\(\) is able to get a multibyte unicode character without problem\,
but pushing a multibyte character back on to the stream causes all
sorts of problems\. From the test cases below\, it looks like ord\(\) is
returning the correct unicode code point \(0xC5\)\, but ungetc\(\) is
interpreting it incorrectly as a byte\. And even after trying to put a
character back on the stream\, the stream is completely broken and only
returns undef\.

I've also tested this with files on disk\, instead of in\-memory
scalars\. The error is the same\.

See also

http&#8203;://stackoverflow\.com/questions/14179751
http&#8203;://perlmonks\.org/?node\_id=1011831

\#\!/usr/local/bin/perl

use 5\.016;
use utf8;
use strict;
use warnings;

use Test&#8203;::More;
my $builder = Test&#8203;::More\->builder;
binmode $builder\->output\,         "&#8203;:encoding\(UTF\-8\)";
binmode $builder\->failure\_output\, "&#8203;:encoding\(UTF\-8\)";
binmode $builder\->todo\_output\,    "&#8203;:encoding\(UTF\-8\)";

binmode\(STDIN\,  '&#8203;:encoding\(utf\-8\)'\);
binmode\(STDOUT\, '&#8203;:encoding\(utf\-8\)'\);
binmode\(STDERR\, '&#8203;:encoding\(utf\-8\)'\);

my $string = qq\[aÅb\];                \# use utf8 makes $string UTF\-8\.
my $fh = IO&#8203;::File\->new\(\);
my $c;

$fh\->open\(\\$string\, '\<&#8203;:encoding\(UTF\-8\)'\);

$c = $fh\->getc\(\);
is\($c\, 'a'\);
$c = $fh\->getc\(\);
is\($c\, 'Å'\);                   \# U\+00C5
$fh\->ungetc\(ord\("Å"\)\);
$c = $fh\->getc\(\);
\{
&nbsp;    local $TODO = "ungetc\(\) doesn't unget it";
&nbsp;    is\($c\, 'Å'\);                   \# U\+00C5
\}
\{
&nbsp;    local $TODO = "The stream is broken at this point\. getc\(\) returns undef\.";
&nbsp;    $c = $fh\->getc\(\);
&nbsp;    is\($c\, 'b'\);                   \# Stream broken at this point\.
\}

done\_testing\(\);

\_\_END\_\_

ok 1
ok 2
not ok 3 \# TODO ungetc\(\) doesn't unget it
\#   Failed \(TODO\) test at unicodify\.pl line 32\.
"\\x\{00c5\}" does not map to utf8 at /usr/local/lib/perl5/5\.16\.2/Test/Builder\.pm line 1759\.
\#          got&#8203;: '\\xC5b'
\#     expected&#8203;: 'Å'
not ok 4 \# TODO The stream is broken at this point\. getc\(\) returns undef\.
\#   Failed \(TODO\) test at unicodify\.pl line 37\.
\#          got&#8203;: undef
\#     expected&#8203;: 'b'
1\.\.4

\_\_END\_\_

$ perl \-v

This is perl 5\, version 16\, subversion 2 \(v5\.16\.2\) built for darwin\-2level

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=high

Site configuration information for perl 5.16.2:

Configured by michael at Wed Nov 21 10:46:56 PST 2012.

Summary of my perl5 (revision 5 version 16 subversion 2) configuration:
   
  Platform:
    osname=darwin, osvers=12.2.0, archname=darwin-2level
    uname='darwin bernard.local 12.2.0 darwin kernel version 12.2.0: sat aug 25 00:48:52 pdt 2012; root:xnu-2050.18.24~1release_x86_64 x86_64 '
    config_args='-de'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-common -DPERL_DARWIN -fno-strict-aliasing -pipe -fstack-protector -I/opt/local/include',
    optimize='-O3',
    cppflags='-fno-common -DPERL_DARWIN -fno-strict-aliasing -pipe -fstack-protector -I/opt/local/include'
    ccversion='', gccversion='4.2.1 Compatible Apple Clang 4.1 ((tags/Apple/clang-421.11.66))', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='env MACOSX_DEPLOYMENT_TARGET=10.3 cc', ldflags =' -fstack-protector -L/opt/local/lib'
    libpth=/opt/local/lib /usr/lib
    libs=-ldbm -ldl -lm -lutil -lc
    perllibs=-ldl -lm -lutil -lc
    libc=, so=dylib, useshrplib=false, libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=bundle, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags=' -bundle -undefined dynamic_lookup -L/opt/local/lib -fstack-protector'

Locally applied patches:
    


@INC for perl 5.16.2:
    /usr/local/lib/perl5/site_perl/5.16.2/darwin-2level
    /usr/local/lib/perl5/site_perl/5.16.2
    /usr/local/lib/perl5/5.16.2/darwin-2level
    /usr/local/lib/perl5/5.16.2
    .


Environment for perl 5.16.2:
    DYLD_LIBRARY_PATH=:/usr/local/pgsql/lib
    HOME=/Users/michael
    LANG=en_CA
    LANGUAGE (unset)
    LC_ALL=C
    LD_LIBRARY_PATH=:/usr/local/pgsql/lib
    LOGDIR (unset)
    PATH=/usr/local/bin:/usr/local/mysql/bin:/usr/local/pgsql/bin:/Users/michael/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/git/bin:/usr/X11/bin:/opt/local/bin:/opt/local/sbin
    PERL_BADLANG (unset)
    SHELL=/bin/bash


```
</details>

**Comment by p5pRT** [comment]
### From @Leont

On Mon\, Jan 7\, 2013 at 6&#8203;:40 AM\, Michael Joyce \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> getc\(\) is able to get a multibyte unicode character without problem\,
> but pushing a multibyte character back on to the stream causes all
> sorts of problems\. From the test cases below\, it looks like ord\(\) is
> returning the correct unicode code point \(0xC5\)\, but ungetc\(\) is
> interpreting it incorrectly as a byte\. And even after trying to put a
> character back on the stream\, the stream is completely broken and only
> returns undef\.
>
> I've also tested this with files on disk\, instead of in\-memory
> scalars\. The error is the same\.
>
> See also
>
> http&#8203;://stackoverflow\.com/questions/14179751
> http&#8203;://perlmonks\.org/?node\_id=1011831
>
> \#\!/usr/local/bin/perl
>
> use 5\.016;
> use utf8;
> use strict;
> use warnings;
>
> use Test&#8203;::More;
> my $builder = Test&#8203;::More\->builder;
> binmode $builder\->output\,         "&#8203;:encoding\(UTF\-8\)";
> binmode $builder\->failure\_output\, "&#8203;:encoding\(UTF\-8\)";
> binmode $builder\->todo\_output\,    "&#8203;:encoding\(UTF\-8\)";
>
> binmode\(STDIN\,  '&#8203;:encoding\(utf\-8\)'\);
> binmode\(STDOUT\, '&#8203;:encoding\(utf\-8\)'\);
> binmode\(STDERR\, '&#8203;:encoding\(utf\-8\)'\);
>
> my $string = qq\[aÅb\];                \# use utf8 makes $string UTF\-8\.
> my $fh = IO&#8203;::File\->new\(\);
> my $c;
>
> $fh\->open\(\\$string\, '\<&#8203;:encoding\(UTF\-8\)'\);
>
> $c = $fh\->getc\(\);
> is\($c\, 'a'\);
> $c = $fh\->getc\(\);
> is\($c\, 'Å'\);                   \# U\+00C5
> $fh\->ungetc\(ord\("Å"\)\);
> $c = $fh\->getc\(\);
> \{
>         local $TODO = "ungetc\(\) doesn't unget it";
>         is\($c\, 'Å'\);                   \# U\+00C5
> \}
> \{
>         local $TODO = "The stream is broken at this point\. getc\(\) returns undef\.";
>         $c = $fh\->getc\(\);
>         is\($c\, 'b'\);                   \# Stream broken at this point\.
> \}
>
> done\_testing\(\);
>
> \_\_END\_\_
>
> ok 1
> ok 2
> not ok 3 \# TODO ungetc\(\) doesn't unget it
> \#   Failed \(TODO\) test at unicodify\.pl line 32\.
> "\\x\{00c5\}" does not map to utf8 at /usr/local/lib/perl5/5\.16\.2/Test/Builder\.pm line 1759\.
> \#          got&#8203;: '\\xC5b'
> \#     expected&#8203;: 'Å'
> not ok 4 \# TODO The stream is broken at this point\. getc\(\) returns undef\.
> \#   Failed \(TODO\) test at unicodify\.pl line 37\.
> \#          got&#8203;: undef
> \#     expected&#8203;: 'b'
> 1\.\.4
>
> \_\_END\_\_

It seems ungetc is \*entirely\* unicode\-unaware \[1\]\. Looks like this can
be fixed easily though\.

Leon

1&#8203;: http&#8203;://perl5\.git\.perl\.org/perl\.git/blob/HEAD&#8203;:/dist/IO/IO\.xs\#l327

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From chansen@cpan.org

On Wed Jan 09 13&#8203;:37&#8203;:10 2013\, LeonT wrote&#8203;:
> It seems ungetc is \*entirely\* unicode\-unaware \[1\]\. Looks like this can
> be fixed easily though\.
> 

Perhaps something like this&#8203;:

\#ifdef PerlIO
&nbsp;        UV v;

&nbsp;        if \(\(SvIOK\_notUV\(c\) && SvIV\(c\) \< 0\) || \(SvNOK\(c\) && SvNV\(c\) \< 0\.0\)\)
&nbsp;            croak\("Negative character number in ungetc\(\)"\);

&nbsp;        v = SvUV\(c\);
&nbsp;        if \(v \<= 0x7F || \(v \<= 0xFF && \!PerlIO\_isutf8\(handle\)\)\)
&nbsp;            RETVAL = PerlIO\_ungetc\(handle\, \(int\)v\);
&nbsp;        else \{
&nbsp;            U8 buf\[UTF8\_MAXBYTES \+ 1\]\, \*end;
&nbsp;            Size\_t len;

&nbsp;            if \(\!PerlIO\_isutf8\(handle\)\)
&nbsp;                croak\("Wide character number in ungetc\(\)"\);

&nbsp;            end = uvchr\_to\_utf8\_flags\(buf\, v\, 0\); /\* XXX flags? \*/
&nbsp;            len = end \- buf;
&nbsp;            if \(PerlIO\_unread\(handle\, &buf\, len\) == len\)
&nbsp;                XSRETURN\_UV\(v\);
&nbsp;            else
&nbsp;                RETVAL = EOF;
&nbsp;        \}
\#else
&nbsp;        RETVAL = ungetc\(\(int\)SvIV\(c\)\, handle\);
\#endif

Returning \-1 when ungetc\(\) is unsuccessful feels wrong\, wouldn't it be more perlish to return undef?

\-\-
chansen

**Comment by p5pRT** [comment]
### From @khwilliamson

Fixed by commit 10e621bc35cb48b15b69b5a57242ff004f7455dc
\-\- 
Karl Williamson

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'open' to 'resolved'
